rfd/0229a-scoped-join-tokens.md

```diff
@@ -0,0 +1,331 @@
++  // provisioning.
++  google.protobuf.Timestamp used_at = 1;
++  // The timestamp representing when a oneshot token should no longer be available
++  // for idempotent retries.
++  google.protobuf.Timestamp reusable_until = 2;
++  // The public key of the host that used the token.
++  bytes used_by_public_key = 3;
++}
++
++// The status of a scoped token.
++message ScopedTokenStatus {
++  // The usage status of the token.
++  oneof usage {
++    // The usage status of a oneshot token.
++    OneshotStatus oneshot = 1;
++  }
++  // The secret that must be provided along with the token's name in order to
++  // be used.
++  string secret = 2;
+```
+
+The `usage` field of the scoped token status is a `oneof` in order to more
+easily support other usage modes that may be added in the future (e.g. max
+usage limits).
+
+The `secret` field represents a departure from the existing token flow in that
+the token's name is no longer the secret. This is similar to how the bounded
+keypair join method manages the registration secret. The `token` join method
+will require specifying both the token's name and secret when using scoped
+tokens.
+
+### Provisioning
+
+Scoped resources will be provisioned by extending the existing `Join` RPC to
+support scoped tokens. When the join method is `token` the auth server will
+query both the existing `ProvisionTokenV2` and the new `ScopedToken` resources
+using the provided token name. Conflicting token names between scoped and
+unscoped tokens will result in an error and it will be up to the administrator
+to resolve the ambiguity. Either by deleting one of the conflicting tokens or
+creating a new token with a different name.
+
+A new `token_secret` field will be included in the `TokenInit` message when
+joining with a scoped token. This secret must match the secret found in the
+retrieved token in order to proceed.
+
+The `assigned_scope` of the `ScopedToken` will be assigned as the scope of the
+resulting node. It will also be attached as metadata to the node's host
+certificates under the `AgentScope` field so that scope related access controls
+can be performed against the agent's identity rather than just restricting
+clients' access to the agent itself.
+
+### Automatic labels for SSH nodes
+
+Scoped tokens should also support automatic assignment of immutable labels to
+any provisioned Teleport SSH agents. This document only proposes support for
+Teleport SSH agents as applying them to other scoped Teleport services (i.e.
+Application, Database, etc.) introduces significant complexity. The
+`ssh_labels` field of a `ScopedToken` will be encoded into the resulting host
+certificates in much the same way as the `assigned_scope`. These labels will be
+extracted while registering the inventory control stream and applied as a new
+`token_labels` field stored in `ServerSpecV2`. Future heartbeats will verify
+`token_labels` just as they do for static labels and command labels.
```

Comment:
Are `token_labels` considered in all of the same places that existing labels are?